src/main/java/se/liu/ida/hefquin/engine/queryproc/impl/poptimizer/simple/CardinalityBasedGreedyJoinPlanOptimizerImpl.java

```diff
@@ -0,0 +1,350 @@
+        protected final List<PhysicalPlan> subplans;
+
+        public GreedyConstructionAlgorithm( final List<PhysicalPlan> subplans ) {
+            this.subplans = subplans;
+        }
+
+        @Override
+        public PhysicalPlan getResultingPlan() throws PhysicalOptimizationException
+        {
+            // Create a list of RequestMemberPair objects for all request operators inside the given list of subplans,
+            // where the list of these objects is ordered in the order in which
+            // the request operators can be found by a depth-first traversal of the subplans.
+            final List<RequestMemberPair> flattenRequestMemPairs = new ArrayList<>();
+            for ( final PhysicalPlan subplan : subplans ) {
+                flattenRequestMemPairs.addAll( createRequestMemPairsFromSourceAssignment(subplan) );
+            }
+
+            // Next, get a list of cardinalities by performing cardinality requests with above list of RequestMemberPair objects.
+            final CardinalityResponse[] resps;
+            try {
+                resps = FederationAccessUtils.performCardinalityRequests(fedAccessMgr, flattenRequestMemPairs.toArray(new RequestMemberPair[0]));
+            } catch (final FederationAccessException e) {
+                throw new PhysicalOptimizationException("Issuing a cardinality request caused an exception.", e);
+            }
+
+            // Then, annotate each subplan (in the 'subplans') with all statistics relevant for the join planning
+            // ( including cardinality, number of access, and a list of relevant federation members )
+            final List<PhysicalPlanWithStatistics> subPlansWithStatistics = associateCardWithSubPlans( flattenRequestMemPairs, resps);
+
+            // To build a left-deep query plan, first select the subplan with the lowest estimated cardinality
+            PhysicalPlanWithStatistics nextSelectedSubPlan = chooseFirstSubPlan(subPlansWithStatistics);
+            subPlansWithStatistics.remove(nextSelectedSubPlan);
+
+            PhysicalPlanWithStatistics currentJoinPlan = nextSelectedSubPlan;
+
+            // The nextSubPlan is selected from subplans that have join variables with selected subplans.
+            // To achieve this, we implement a custom comparator and create a PriorityQueue to store candidate subPlans in cardinality order.
+            final Comparator<PhysicalPlanWithStatistics> orderBasedOnCard = new Comparator<PhysicalPlanWithStatistics>() {
+                @Override
+                public int compare( PhysicalPlanWithStatistics o1, PhysicalPlanWithStatistics o2 ) {
+                    return Integer.compare( o1.getCardinality(), o2.getCardinality() );
+                }
+            };
+            final PriorityQueue<PhysicalPlanWithStatistics> orderedCandidateSubPlans = new PriorityQueue<>(orderBasedOnCard);
+
+            for ( int i = 1; i < subplans.size(); i ++ ){
```

Comment:
This has confused me a lot (Why is the iteration starting from index 1 here? is it because a first subplan was already picked? But that subplan would not necessarily be the first subplan (with index 0)!).

So, I had to think twice about this. I understand now that this is correct, but a much more obvious and less confusing thing would be to have a while loop here, with the while condition that at least one of the following lists is not empty yet: `orderedCandidateSubPlans` and `orderedCandidateSubPlans`.
```suggestion
            while ( ! orderedCandidateSubPlans.isEmpty() && ! orderedCandidateSubPlans.isEmpty() ) {
```

From the code readability perspective, I would find it more natural to

Reply:
It does make more sense! Just for confirmation, do you mean to refer to these lists: 'orderedCandidateSubPlans' and 'subPlansWithStatistics'?

Reply:
Yes. These are the two lists that I meant. Sorry for the typo! (must have been a copy & paste mistake)